crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs

```diff
@@ -0,0 +1,972 @@
+            }
+        };
+        if self.have_new_data() || self.inner.share_mut(update_last_sync)? {
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// Record all timed out rounds into metrics
+    fn record_timeouts(&self) {
+        self.inner
+            .share_mut(|i, _| {
+                for cert in i.round_map.take_timed_out_rounds() {
+                    let metric = MetricInfo {
+                        error: true,
+                        last_sync: std::time::Instant::now(),
+                    };
+                    i.metrics.insert(cert, metric);
+                }
+                Ok(())
+            })
+            .ok();
+    }
+}
+
+/// Calculates the arcset used during local sync. This arcset determines the
+/// minimum set of ops to be spread across all local agents in order to reach
+/// local consistency
+fn local_sync_arcset(arcs: &[ArcInterval]) -> DhtArcSet {
+    arcs.iter()
+        .enumerate()
+        // For each agent's arc,
+        .map(|(i, arc_i)| {
+            // find the union of all arcs *other* than this one,
+            let other_arcset = arcs
+                .iter()
+                .enumerate()
+                .filter_map(|(j, arc_j)| {
+                    if i == j {
+                        None
+                    } else {
+                        Some(DhtArcSet::from(arc_j))
+                    }
+                })
+                .fold(DhtArcSet::new_empty(), |a, b| DhtArcSet::union(&a, &b));
+            // and return the intersection of this arc with the union of the others.
+            DhtArcSet::from(arc_i).intersection(&other_arcset)
+        })
+        // and take the union of all of the intersections
+        .fold(DhtArcSet::new_empty(), |a, b| DhtArcSet::union(&a, &b))
+}
+
+impl RoundState {
+    fn increment_sent_ops_blooms(&mut self) -> u8 {
+        self.num_sent_ops_blooms += 1;
+        self.num_sent_ops_blooms
+    }
+}
+
+/// Time range from now into the past.
+/// Start must be < end.
+fn time_range(start: Duration, end: Duration) -> Range<u64> {
+    let now = SystemTime::now();
+    let start = now
+        .checked_sub(start)
+        .and_then(|t| t.duration_since(SystemTime::UNIX_EPOCH).ok())
+        .map(|t| t.as_millis() as u64)
+        .unwrap_or(0);
+
+    let end = now
+        .checked_sub(end)
+        .and_then(|t| t.duration_since(SystemTime::UNIX_EPOCH).ok())
+        .map(|t| t.as_millis() as u64)
+        .unwrap_or(0);
+
+    start..end
+}
+
+kitsune_p2p_types::write_codec_enum! {
+    /// SimpleBloom Gossip Wire Protocol Codec
+    codec ShardedGossipWire {
+        /// Initiate a round of gossip with a remote node
+        Initiate(0x10) {
+            /// The list of arc intervals (equivalent to a [`DhtArcSet`])
+            /// for all local agents
+            intervals.0: Vec<ArcInterval>,
+            /// A random number to resolve concurrent initiates.
+            id.1: u8,
```

Comment:
the odds of them both picking the same number are not astronomical with 256 options... but I suppose it's not a big deal whatever happens in that case.

Reply:
Yeh for sure, I was just trying to save space. I don't think it's a huge deal if they both fail occasionally but can make this a u32 if you reckon it's worth it?

Reply:
nah